In [56]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from utils import bootcampviztools as bt
from utils import toolbox_ML as tl

Link de Kaggle modelo predictivo: https://www.kaggle.com/datasets/harlfoxem/housesalesprediction/data

Fuente oficial de los datos: https://geodacenter.github.io/data-and-lab/KingCounty-HouseSales2015/

#### OBJETIVO DE DE ESTE PROYECTO

El objetivo de es proyecto es predecir el precio de los pisos teniendo en cuenta una serie de características como el número de camas, año de construcción, número de plantas, ... Los datos con los que se entrenarán al modelo han sido obtenidos de un dataset de **Kaggle**. Los datos contienen información sobre casas vendidas del condado de King, Washington durante 2014 - 2015. El modelo propuesto consistirá en un ``modelo de regresión supervisado``.

#### Carga de los datos

In [57]:
df = pd.read_csv('./data/kc_house_data.csv')

In [58]:
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

In [59]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  object 
 2   price          21613 non-null  float64
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21613 non-null  float64
 5   sqft_living    21613 non-null  int64  
 6   sqft_lot       21613 non-null  int64  
 7   floors         21613 non-null  float64
 8   waterfront     21613 non-null  int64  
 9   view           21613 non-null  int64  
 10  condition      21613 non-null  int64  
 11  grade          21613 non-null  int64  
 12  sqft_above     21613 non-null  int64  
 13  sqft_basement  21613 non-null  int64  
 14  yr_built       21613 non-null  int64  
 15  yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  lat            21613 non-null  float64
 18  long  

El df no contiene ningún valor nulo

In [60]:
df.describe().T

count          mean           std           min  \
id             21613.0  4.580302e+09  2.876566e+09  1.000102e+06   
price          21613.0  5.400881e+05  3.671272e+05  7.500000e+04   
bedrooms       21613.0  3.370842e+00  9.300618e-01  0.000000e+00   
bathrooms      21613.0  2.114757e+00  7.701632e-01  0.000000e+00   
sqft_living    21613.0  2.079900e+03  9.184409e+02  2.900000e+02   
sqft_lot       21613.0  1.510697e+04  4.142051e+04  5.200000e+02   
floors         21613.0  1.494309e+00  5.399889e-01  1.000000e+00   
waterfront     21613.0  7.541757e-03  8.651720e-02  0.000000e+00   
view           21613.0  2.343034e-01  7.663176e-01  0.000000e+00   
condition      21613.0  3.409430e+00  6.507430e-01  1.000000e+00   
grade          21613.0  7.656873e+00  1.175459e+00  1.000000e+00   
sqft_above     21613.0  1.788391e+03  8.280910e+02  2.900000e+02   
sqft_basement  21613.0  2.915090e+02  4.425750e+02  0.000000e+00   
yr_built       21613.0  1.971005e+03  2.937341e+01  1.900000e+03   
yr_renovated   21613.0  8.440226e+01  4.016792e+02  0.000000e+00   
zipcode        21613.0  9.807794e+04  5.350503e+01  9.800100e+04   
lat            21613.0  4.756005e+01  1.385637e-01  4.715590e+01   
long           21613.0 -1.222139e+02  1.408283e-01 -1.225190e+02   
sqft_living15  21613.0  1.986552e+03  6.853913e+02  3.990000e+02   
sqft_lot15     21613.0  1.276846e+04  2.730418e+04  6.510000e+02   

                        25%           50%           75%           max  
id             2.123049e+09  3.904930e+09  7.308900e+09  9.900000e+09  
price          3.219500e+05  4.500000e+05  6.450000e+05  7.700000e+06  
bedrooms       3.000000e+00  3.000000e+00  4.000000e+00  3.300000e+01  
bathrooms      1.750000e+00  2.250000e+00  2.500000e+00  8.000000e+00  
sqft_living    1.427000e+03  1.910000e+03  2.550000e+03  1.354000e+04  
sqft_lot       5.040000e+03  7.618000e+03  1.068800e+04  1.651359e+06  
floors         1.000000e+00  1.500000e+00  2.000000e+00  3.500000e+00  
waterfront     0.000000e+00  0.000000e+00  0.000000e+00  1.000000e+00  
view           0.000000e+00  0.000000e+00  0.000000e+00  4.000000e+00  
condition      3.000000e+00  3.000000e+00  4.000000e+00  5.000000e+00  
grade          7.000000e+00  7.000000e+00  8.000000e+00  1.300000e+01  
sqft_above     1.190000e+03  1.560000e+03  2.210000e+03  9.410000e+03  
sqft_basement  0.000000e+00  0.000000e+00  5.600000e+02  4.820000e+03  
yr_built       1.951000e+03  1.975000e+03  1.997000e+03  2.015000e+03  
yr_renovated   0.000000e+00  0.000000e+00  0.000000e+00  2.015000e+03  
zipcode        9.803300e+04  9.806500e+04  9.811800e+04  9.819900e+04  
lat            4.747100e+01  4.757180e+01  4.767800e+01  4.777760e+01  
long          -1.223280e+02 -1.222300e+02 -1.221250e+02 -1.213150e+02  
sqft_living15  1.490000e+03  1.840000e+03  2.360000e+03  6.210000e+03  
sqft_lot15     5.100000e+03  7.620000e+03  1.008300e+04  8.712000e+05

In [61]:
'''Variable target'''

target = 'price'

Tabla con información de las variables 

| Variable        | Tipo    | Descripción | Notas |
|---------------|--------|-------------|-------|
| id            | int64  | Identificador único de la vivienda | Puede no ser relevante para el análisis |
| date          | object | Fecha de venta de la vivienda | Formato de fecha puede requerir conversión |
| price         | float64| Precio de venta de la vivienda en dólares | Variable objetivo para predicción de precios |
| bedrooms      | int64  | Número de habitaciones en la vivienda | Puede incluir dormitorios pequeños |
| bathrooms     | float64| Número de baños en la vivienda | Considerado medios baños como 0.5 (los que no tienen ducha) |
| sqft_living   | int64  | Metros cuadrados de la vivienda habitable | Relacionado con el precio |
| sqft_lot      | int64  | Metros cuadrados del terreno | Incluye jardín y otros espacios exteriores |
| floors        | float64| Número de pisos de la vivienda | Puede ser decimal si hay entrepisos |
| waterfront    | int64  | 1 si la vivienda tiene vista al agua, 0 si no | Variable categórica binaria |
| view          | int64  | Puntuación de la vista de la vivienda (0-4) | 0 indica sin vista, 4 es la mejor vista |
| condition     | int64  | Estado general de la vivienda (1-5) | 1 es malo, 5 es excelente |
| grade         | int64  | Calidad de la construcción y acabados (1-13) | Basado en estándares de construcción |
| sqft_above    | int64  | Metros cuadrados de la parte sobre el suelo | Excluye sótano |
| sqft_basement | int64  | Metros cuadrados del sótano | 0 si no tiene sótano |
| yr_built      | int64  | Año de construcción de la vivienda | Puede influir en el precio y estado |
| yr_renovated  | int64  | Año de la última renovación, 0 si nunca ha sido renovada | Puede afectar el valor de la vivienda |
| zipcode       | int64  | Código postal de la ubicación | Puede utilizarse para análisis geoespacial |
| lat           | float64| Latitud de la vivienda | Coordenada geográfica |
| long          | float64| Longitud de la vivienda | Coordenada geográfica |
| sqft_living15 | int64  | Metros cuadrados promedio de las viviendas cercanas | Indicador del valor del vecindario |
| sqft_lot15    | int64  | Metros cuadrados promedio del terreno en la zona | Puede influir en el precio |

#### Tipo de feature, cardinalidad, porcentaje de nulos y valores únicos

In [62]:
tl.describe_df(df)

columna     tipo  %_nulos  valores_unicos  %_cardinalidad
0              id    int64      0.0           21436           99.18
1            date   object      0.0             372            1.72
2           price  float64      0.0            4028           18.64
3        bedrooms    int64      0.0              13            0.06
4       bathrooms  float64      0.0              30            0.14
5     sqft_living    int64      0.0            1038            4.80
6        sqft_lot    int64      0.0            9782           45.26
7          floors  float64      0.0               6            0.03
8      waterfront    int64      0.0               2            0.01
9            view    int64      0.0               5            0.02
10      condition    int64      0.0               5            0.02
11          grade    int64      0.0              12            0.06
12     sqft_above    int64      0.0             946            4.38
13  sqft_basement    int64      0.0             306            1.42
14       yr_built    int64      0.0             116            0.54
15   yr_renovated    int64      0.0              70            0.32
16        zipcode    int64      0.0              70            0.32
17            lat  float64      0.0            5034           23.29
18           long  float64      0.0             752            3.48
19  sqft_living15    int64      0.0             777            3.60
20     sqft_lot15    int64      0.0            8689           40.20

La columna id no contiene un 100% de la cardinalidad, pero se queda muy próximo. Esto indica que hay algunos índices que están duplicados en el df algo que no es común ya que el id debería de ser un identificador único.

In [63]:
df['id'].duplicated().value_counts()

id
False    21436
True       177
Name: count, dtype: int64

In [64]:
df['id'].duplicated().value_counts(normalize=True)

id
False    0.99181
True     0.00819
Name: proportion, dtype: float64

Las instancias con 'id' duplicado suponen menos de un 1%

In [65]:
df.loc[df['id'].duplicated(keep = False)].sort_values(by = 'id', ascending=False).head(20)

id             date      price  bedrooms  bathrooms  \
1085   9834200885  20140717T000000   360000.0         4       2.50   
1086   9834200885  20150420T000000   550000.0         4       2.50   
15199  9834200305  20140716T000000   350000.0         3       1.00   
15200  9834200305  20150210T000000   615000.0         3       1.00   
6345   9828200460  20140627T000000   260000.0         2       1.00   
6346   9828200460  20150106T000000   430000.0         2       1.00   
4922   9809000020  20140513T000000  1895000.0         5       2.25   
4923   9809000020  20150313T000000  1940000.0         5       2.25   
2493   9407110710  20141107T000000   195000.0         3       1.75   
2494   9407110710  20150226T000000   322000.0         3       1.75   
4342   9353300600  20140624T000000   348500.0         3       1.50   
4343   9353300600  20150326T000000   370000.0         3       1.50   
16672  9250900104  20150410T000000   496000.0         5       1.75   
16671  9250900104  20141110T000000   300000.0         5       1.75   
12832  9238500040  20140624T000000   400000.0         3       2.50   
12833  9238500040  20150210T000000   599000.0         3       2.50   
6371   9222400605  20141115T000000   842500.0         5       4.00   
6372   9222400605  20150411T000000   850000.0         5       4.00   
13024  9211500620  20150428T000000   305000.0         3       2.25   
13023  9211500620  20141008T000000   182700.0         3       2.25   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
1085          2080      4080     1.0           0     0  ...      7   
1086          2080      4080     1.0           0     0  ...      7   
15199         1790      3876     1.5           0     0  ...      7   
15200         1790      3876     1.5           0     0  ...      7   
6345           700      4800     1.0           0     0  ...      7   
6346           700      4800     1.0           0     0  ...      7   
4922          3120     16672     2.0           0     0  ...      9   
4923          3120     16672     2.0           0     0  ...      9   
2493          1510      8400     1.0           0     0  ...      7   
2494          1510      8400     1.0           0     0  ...      7   
4342          1360     10726     1.0           0     0  ...      7   
4343          1360     10726     1.0           0     0  ...      7   
16672         2110      8500     1.0           0     0  ...      7   
16671         2110      8500     1.0           0     0  ...      7   
12832         2970     23100     1.0           0     0  ...      7   
12833         2970     23100     1.0           0     0  ...      7   
6371          2980      4500     1.5           0     0  ...      7   
6372          2980      4500     1.5           0     0  ...      7   
13024         1740      6650     1.0           0     0  ...      7   
13023         1740      6650     1.0           0     0  ...      7   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
1085         1040           1040      1962             0    98144  47.5720   
1086         1040           1040      1962             0    98144  47.5720   
15199        1090            700      1904             0    98144  47.5750   
15200        1090            700      1904             0    98144  47.5750   
6345          700              0      1922             0    98122  47.6147   
6346          700              0      1922             0    98122  47.6147   
4922         3120              0      1969             0    98004  47.6458   
4923         3120              0      1969             0    98004  47.6458   
2493          980            530      1979             0    98045  47.4476   
2494          980            530      1979             0    98045  47.4476   
4342         1360              0      1966             0    98059  47.4948   
4343         1360              0      1966             0    98059  47.4948   
16672        1100           1010      1962             0   

Viendo los valores que tienen un id duplicado se observa que los únicos valores que cambian son el precio de venta de la casa y la fecha de venta. Esto indica que hay casas que han sido vendidas más de una vez.

In [66]:
df['ventas'] = df.groupby('id')['id'].transform('count')
df[df['ventas'] > 1]
df.ventas.value_counts()

ventas
1    21260
2      350
3        3
Name: count, dtype: int64

In [67]:
df.loc[df['ventas']==3]

id             date     price  bedrooms  bathrooms  sqft_living  \
17602  795000620  20140924T000000  115000.0         3        1.0         1080   
17603  795000620  20141215T000000  124000.0         3        1.0         1080   
17604  795000620  20150311T000000  157000.0         3        1.0         1080   

       sqft_lot  floors  waterfront  view  ...  sqft_above  sqft_basement  \
17602      6250     1.0           0     0  ...        1080              0   
17603      6250     1.0           0     0  ...        1080              0   
17604      6250     1.0           0     0  ...        1080              0   

       yr_built  yr_renovated  zipcode      lat    long  sqft_living15  \
17602      1950             0    98168  47.5045 -122.33           1070   
17603      1950             0    98168  47.5045 -122.33           1070   
17604      1950             0    98168  47.5045 -122.33           1070   

       sqft_lot15  ventas  
17602        6250       3  
17603        6250       3  
17604        6250       3  

[3 rows x 22 columns]

Hay un total de 350 casas que han sido vendidas 2 veces y 3 casas han sido vendidas 3 veces.

Como nuestro objetivo principal consiste en predecir el precio de una casa de manera inicial nos quedaremos con el primer registro de la venta de la casa y eliminaremos el resto ya que las futuras ventas han sido condicionadas por el precio anterior de la misma.

In [68]:
df = df.drop_duplicates(subset=['id'], keep='first')

In [69]:
df.ventas.value_counts()

ventas
1    21260
2      175
3        1
Name: count, dtype: int64

In [70]:
df.loc[df['ventas']==3]

id             date     price  bedrooms  bathrooms  sqft_living  \
17602  795000620  20140924T000000  115000.0         3        1.0         1080   

       sqft_lot  floors  waterfront  view  ...  sqft_above  sqft_basement  \
17602      6250     1.0           0     0  ...        1080              0   

       yr_built  yr_renovated  zipcode      lat    long  sqft_living15  \
17602      1950             0    98168  47.5045 -122.33           1070   

       sqft_lot15  ventas  
17602        6250       3  

[1 rows x 22 columns]

In [71]:
'''Cambiar formato de la fecha de venta'''
df = df.rename({'date':'sale_date'}, axis = 1)
df['sale_date'] = pd.to_datetime(df['sale_date'], yearfirst=True)

In [72]:
df.describe().T

count                           mean                  min  \
id             21436.0              4580765328.177878            1000102.0   
sale_date        21436  2014-10-28 05:22:26.893076992  2014-05-02 00:00:00   
price          21436.0                   540529.28718              75000.0   
bedrooms       21436.0                       3.371571                  0.0   
bathrooms      21436.0                       2.117349                  0.0   
sqft_living    21436.0                    2082.704936                290.0   
sqft_lot       21436.0                   15135.637852                520.0   
floors         21436.0                       1.496198                  1.0   
waterfront     21436.0                       0.007604                  0.0   
view           21436.0                       0.235118                  0.0   
condition      21436.0                       3.410384                  1.0   
grade          21436.0                       7.661737                  1.0   
sqft_above     21436.0                     1790.96044                290.0   
sqft_basement  21436.0                     291.744495                  0.0   
yr_built       21436.0                    1971.098433               1900.0   
yr_renovated   21436.0                        84.7298                  0.0   
zipcode        21436.0                   98077.862288              98001.0   
lat            21436.0                      47.560156              47.1559   
long           21436.0                    -122.213697             -122.519   
sqft_living15  21436.0                    1988.314378                399.0   
sqft_lot15     21436.0                    12785.96128                651.0   
ventas         21436.0                       1.008257                  1.0   

                               25%                  50%                  75%  \
id                   2123700078.75         3904921185.0         7308675062.5   
sale_date      2014-07-21 00:00:00  2014-10-15 00:00:00  2015-02-13 00:00:00   
price                     322150.0             450000.0             645000.0   
bedrooms                       3.0                  3.0                  4.0   
bathrooms                     1.75                 2.25                  2.5   
sqft_living                 1430.0               1920.0               2550.0   
sqft_lot                    5040.0               7614.0             10696.25   
floors                         1.0                  1.5                  2.0   
waterfront                     0.0                  0.0                  0.0   
view                           0.0                  0.0                  0.0   
condition                      3.0                  3.0                  4.0   
grade                          7.0                  7.0                  8.0   
sqft_above                  1200.0               1560.0               2220.0   
sqft_basement                  0.0                  0.0                560.0   
yr_built                    1952.0               1975.0               1997.0   
yr_renovated                   0.0                  0.0                  0.0   
zipcode                    98033.0              98065.0              98117.0   
lat                        47.4711               47.572               47.678   
long                      -122.328              -122.23             -122.124   
sqft_living15               1490.0               1840.0               2370.0   
sqft_lot15                  5100.0               7620.0             10087.25   
ventas                         1.0                  1.0                  1.0   

                               max               std  
id                    9900000190.0  2876589633.67309  
sale_date      2015-05-27 00:00:00               NaN  
price                    7700000.0     367689.296471  
bedrooms                      33.0          0.929205  
bathrooms                      8.0          0.769913  
sqft_living                13540.0        919.14646

Tambien se puede ver que el valor máximo de `bedrooms` es 33, mientras que el 75% de los valores se encuentran por debajo de 4. Veamos cuantas viviendas tienen esa cantidad de habitaciones.

In [73]:
df.loc[df['bedrooms'] == 33]

id  sale_date     price  bedrooms  bathrooms  sqft_living  \
15870  2402100895 2014-06-25  640000.0        33       1.75         1620   

       sqft_lot  floors  waterfront  view  ...  sqft_above  sqft_basement  \
15870      6000     1.0           0     0  ...        1040            580   

       yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
15870      1947             0    98103  47.6878 -122.331           1330   

       sqft_lot15  ventas  
15870        4700       1  

[1 rows x 22 columns]

Como solo es una vivienda, eliminaremos esta instancia.

In [74]:
bedrooms_33 = df[df['bedrooms'] == 33].index
df = df.drop(bedrooms_33)

In [75]:
df = df.set_index('sale_date')

In [76]:
df_limpio = df.to_csv('./data/df_limpio.csv')